Add batch.kubernetes.io to labels created in the Job controller.

kannon92 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind cleanup
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
In https://github.com/kubernetes/kubernetes/issues/113714, there was a suggestion to use Kubernetes prefix labels for any label created by Kubernetes.  Unprefixed labels are reserved for end-users.  This PR adds a prefix to job-name and controller-uid but keeps the existing label also.    
#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #113714 

#### Special notes for your reviewer:
I am unsure how to state that the legacy behavior will be depreciated eventually.  I think we would want this to set in production for a few releases.
#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Pods owned by a Job will now use the labels `batch.kubernetes.io/job-name` and `batch.kubernetes.io/controller-uid`.
The legacy labels `job-name` and `controller-uid` are still added for compatibility.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

kannon92 committed:
Add batch.kubernetes.io to labels created in the Job controller.

sftim commented:
How about this changelog entry:
```markdown
Fixed a bug where Kubernetes used unprefixed labels (and defined matching selectors) for Jobs and the
objects they managed.
Pods owned by a Job will now use the labels `batch.kubernetes.io/job-name` and `batch.kubernetes.io/controller-uid`.
The legacy labels `job-name` and `controller-uid` are still accepted for compatibility, but are never set.
```

sftim commented:
If we make this change, what happens to existing pods for Jobs? Do they still get selected?

I wonder if either:
- we end up making twice as many queries to select Pods, because we have to OR together the outcomes of two different selectors
- we forget about the old Jobs

Is there a test case that shows things are OK here?

kannon92 commented:
> If we make this change, what happens to existing pods for Jobs? Do they still get selected?
> 
> I wonder if either:
> 
> * we end up making twice as many queries to select Pods, because we have to OR together the outcomes of two different selectors
> * we forget about the old Jobs
> 
> Is there a test case that shows things are OK here?

I'm still learning the test plan for Kubernetes.  Would this test case be considered a e2e or an integration test?  I added test cases that add both labels.  

I do not have any test case for your request but I can add them once I figure out what type of test that would be.

sftim commented on staging/src/k8s.io/api/batch/v1/types.go:
```diff
@@ -35,6 +35,10 @@ const (
 	// always track jobs using finalizers.
 	JobTrackingFinalizer = "batch.kubernetes.io/job-tracking"
+	// The Job labels will use batch.kubernetes.io as a prefix for all labels
+	// Currently the job controller will set job-name and controller-uid with this label and also sets the job-name
```
Watch out for “Currently” in code; this can end up ageing. It's better to document behavior without that word, and maybe also write something like “Historically, the Job controller used the unprefixed labels (a bug), and Kubernetes continues to recognize those unprefixed labels for compatbility”.

sftim commented:
I'm talking about a test case that shows that things stay working through a control plane upgrade. Not sure what we call those.

kannon92 commented:
I opened up https://github.com/kubernetes/kubernetes/pull/115111 to address some low coverage around the registry job.  I wanted to add some more test cases around the labels but it seems that there is a lot of missing code around this here.  

Figured I would open up a separate PR for that.

kannon92 commented:
@sftim So I've starting looking more into the job controller code and I see that we only use the Selectors for getting pods:

```
func (jm *Controller) getPodsForJob(ctx context.Context, j *batch.Job) ([]*v1.Pod, error) {
	selector, err := metav1.LabelSelectorAsSelector(j.Spec.Selector)
	if err != nil {
		return nil, fmt.Errorf("couldn't convert Job selector: %v", err)
	}
```
Since there is a general concern about version skew and not losing pods via an upgrade, what do you think about not changing the behavior for setting the selector ie leave it in the legacy path for this first round?  

I think in general we can start adding the correct labels in Labels and in the PodTemplate.  And eventually when we phase out the legacy label we can assume that the n-1 release has labels with the non-buggy path.  

I tried to come up with a test case but we don't allow users to modify these selectors and any manual selector must match exactly what we would be geting manually.  So it is not really easy to come up with a mocked upgrade path.

kannon92 commented:
/retest

kannon92 commented:
Update on this PR.  I got someone to review https://github.com/kubernetes/kubernetes/pull/115111 so I am hoping to get those tests merged.  After that I will update this PR and add some more test cases.

We are still looking for someone that has approval rights for https://github.com/kubernetes/kubernetes/pull/115111.

kannon92 commented:
/retest

alculquicondor commented:
/assign

soltysh commented on pkg/apis/batch/types.go:
```diff
@@ -33,6 +33,9 @@ import (
 const JobTrackingFinalizer = "batch.kubernetes.io/job-tracking"
 
+// We will use batch.kubernetes.io/ as a prefix for all labels
+const JobLabelPrefix = "batch.kubernetes.io/"
```
Nit: `BatchLabelPrefix` seems more appropriate

thockin replied:
"batch" is the pkg name, so maybe just "LabelPrefix" ?

kannon92 replied:
Updated everywhere. Good call on the fact that this could be a constant.

soltysh commented on pkg/registry/batch/job/strategy.go:
```diff
@@ -211,23 +211,30 @@ func generateSelector(obj *batch.Job) {
-		// we will reject in validation.
-	} else {
-		obj.Spec.Template.Labels["job-name"] = string(obj.ObjectMeta.Name)
+	jobNameLabels := []string{"job-name", batch.JobLabelPrefix + "job-name"}
```
I'd move that as a global variable, such that we don't create it on every call.

soltysh commented on pkg/registry/batch/job/strategy.go:
```diff
@@ -211,23 +211,30 @@ func generateSelector(obj *batch.Job) {
-		// we will reject in validation.
-	} else {
-		obj.Spec.Template.Labels["controller-uid"] = string(obj.ObjectMeta.UID)
+	controllerUidLabels := []string{"controller-uid", batch.JobLabelPrefix + "controller-uid"}
```
Ditto.

kannon92 replied:
Done.

soltysh commented on pkg/registry/batch/job/strategy.go:
```diff
@@ -236,8 +243,11 @@ func generateSelector(obj *batch.Job) {
-	if _, found := obj.Spec.Selector.MatchLabels["controller-uid"]; !found {
-		obj.Spec.Selector.MatchLabels["controller-uid"] = string(obj.ObjectMeta.UID)
+
+	for _, value := range []string{"controller-uid"} {
```
Don't change this, yet, but add here a TODO, that we'll want to change that in the future.

alculquicondor replied:
why can't we change this to use just the new label?

Anyone that is relying on this selector (for example, the job controller itself) should be able to use it without caring what the labels are.

kannon92 replied:
@sftim brought up a possible bug if there is an upgrade.

1.27 behavior:

a) Legacy labels are present
b) Legacy labels (controller-uid) is used as selector.

Upgrade of Kubernetes (1.28)

New labels would be added for new jobs and the new label would be used for selector.  

I don't really know if upgrades would change the selector and I was being on the conservative side so we can make sure that both labels are present so we could switch to use new selector once the upgrade paths that we can support have both labels.

alculquicondor replied:
> New labels would be added for new jobs and the new label would be used for selector.

Exactly, already existing jobs wouldn't change, as this is part of the creation strategy, not update.

kannon92 replied:
https://github.com/kubernetes/kubernetes/issues/113714#issuecomment-1400581581 was the discussion that led me to implement just using the old label selector.

So you are saying that upgrades should keep the old labels and I don't think we allow updates to the LabelSelector?

thockin replied:
I mistakenly assumed that selector was hardcoded in the controller, but it's not.

soltysh reviewed (commented):
Minor nits, you'll need to get api-review on this PR, though.

soltysh commented:
/assign @thockin 
for api-approval, since he started this endeavor :) 
/triage accepted
/priority important-longterm

thockin commented on staging/src/k8s.io/api/batch/v1/types.go:
```diff
@@ -35,6 +35,10 @@ const (
+	// The Job labels will use batch.kubernetes.io as a prefix for all labels
+	// Historically the job controller uses unprefixed labels for job-name and controller-uid and
+	// Kubernetes continutes to recognize those unprefixed labels for consistency.
+	JobLabelPrefix = "batch.kubernetes.io/"
```
Shouldn't we define these sorts of cross-version constants in `staging/src/k8s.io/api/batch` rather than `.../v1` ?  I know it's not exactly common, but having the same const defined twice feels like a smell to me.

alculquicondor replied:
this is a recurring topic of debates in API reviews :)

Someone writing a controller should use the versioned packages. So we actually should have them here.

thockin replied:
My point was that it seems unlikely that such a key woyuld change across API versions.

I'll resolve this and discuss with other API reviewers async.

alculquicondor commented on pkg/apis/batch/validation/validation.go:
```diff
@@ -99,12 +105,18 @@ func ValidateGeneratedSelector(obj *batch.Job) field.ErrorList {
 	// have placed certain labels on the pod, but this could have failed if
 	// the user added coflicting labels.  Validate that the expected
 	// generated ones are there.
+	allErrs = append(allErrs, apivalidation.ValidateHasLabel(obj.Spec.Template.ObjectMeta, field.NewPath("spec").Child("template").Child("metadata"), controllerUid, string(obj.UID))...)
```
This change is adding extra validation, which is generally hard.

Note that this function is called during JobUpdates. Then every Job update would fail for old jobs.

In general, we need to allow this validation to pass if it was passing before the update. See this guide for a similar change when adding enums https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-enum-value-in-existing-field

kannon92 replied:
I read the guide and I am still a bit confused on this.

So for this release, should I just remove the validation with the new labels? Until the labels are present in a previous release I can't really validate them?

alculquicondor commented on pkg/apis/batch/validation/validation.go:
```diff
@@ -99,12 +105,18 @@ func ValidateGeneratedSelector(obj *batch.Job) field.ErrorList {
+	expectedLabels[controllerUidLegacy] = string(obj.UID)
+	expectedLabels[jobNameLegacy] = string(obj.Name)
+	expectedLabels[controllerUid] = string(obj.UID)
+	expectedLabels[jobName] = string(obj.Name)
```
For that change to be backwards compatible, we need to only validate the Job name with the label value rules only when using a generated selector.

thockin replied:
If you have a Job with a long name (and not manual selector) you will try to stick that long name into the label value.  Given that this is happening in validation, AFTER label validation, you will end up inserting corrupted (too long) labels and (I think) storing them!

The next update should fail validation in a very confusing way.

If you properly set these labels in PrepareForCreate() or even in pkg/apis/batch/v1/defaults.go, then validation would fail - still with a bad error, but at least it would fail.  You could do better by checking the `manualSelector` flag and then failing if the name is too long, with a good error.

Please also add a testcase of creating a Job with a too-long name.

kannon92 replied:
Do you think this can also be done as a followup? 

I don't think this is related to adding these labels as we aren't checking the max length of label plus name.

thockin replied:
sure, this is not a new failure mode - it exists today.

Can you please file another issue to capture? 

Thanks

kannon92 replied:
https://github.com/kubernetes/kubernetes/pull/116146

alculquicondor commented on pkg/apis/batch/validation/validation.go:
```diff
@@ -90,6 +90,12 @@ func ValidateGeneratedSelector(obj *batch.Job) field.ErrorList {
+	controllerUidLegacy := "controller-uid"
+	jobNameLegacy := "job-name"
+	// Use Label Prefix for controller-uid and job name
+	controllerUid := batch.JobLabelPrefix + "controller-uid"
```
they could be package constants themselves

kannon92 replied:
done.

kannon92 commented on pkg/registry/batch/job/strategy.go:
```diff
@@ -211,23 +211,30 @@ func generateSelector(obj *batch.Job) {
+	jobNameLabels := []string{"job-name", batch.JobLabelPrefix + "job-name"}
+	for _, value := range jobNameLabels {
+		_, found := obj.Spec.Template.Labels[value]
+		if found {
```
Does this comment explain it?

```
	// If selector generation was requested, then expected labels must be
	// present on pod template, and must match job's uid and name.  The
	// generated (not-manual) selectors/labels ensure no overlap with other
	// controllers.  The manual mode allows orphaning, adoption,
	// backward-compatibility, and experimentation with new
	// labeling/selection schemes.  Automatic selector generation should
	// have placed certain labels on the pod, but this could have failed if
	// the user added coflicting labels.  Validate that the expected
	// generated ones are there.
```

I think that validation requires that you must have these labels correspond to Name and ControllerUId so not entirely sure what we gain from this.

thockin replied:
So let me see if I get this right (not super familair with Job):

If the user wants a manual selector, they say so and then put whatever value they want in the selector, right?  This function appears to only be called if they didn't set the `manualSelector` field.

So the below comment "User asked us to not automatically generate a selector" is wrong - this whole function is only called when the user DID ask us to generate a selector.  Or do I still misunderstand?

alculquicondor replied:
It should say: "User asked us to automatically generate a selector, but set manual labels, maybe conflicting"

kannon92 replied:
I didn't update this comment in this PR so I opened up https://github.com/kubernetes/kubernetes/pull/116215 where we can fix that.

alculquicondor replied:
It's ok to fix such small things in a PR that is doing something else. But up to you.

kannon92 replied:
Good to know! I usually like to follow the approach of small PRs and keeping them focused on a particular thing.  I didn't really follow my own policy in this one so I've been trying to push things unrelated to this PR to another one.  Mostly so reviews can be detailed and easy to get through.  

I also think I want to look a bit more at manual selectors hence https://github.com/kubernetes/kubernetes/issues/116137

kannon92 replied:
#116215 was merged so I think we get resolve this.

kannon92 commented on pkg/registry/batch/job/strategy.go:
```diff
@@ -211,23 +211,30 @@ func generateSelector(obj *batch.Job) {
+	controllerUidLabels := []string{"controller-uid", batch.JobLabelPrefix + "controller-uid"}
+	for _, value := range controllerUidLabels {
+		_, found := obj.Spec.Template.Labels[value]
+		if found {
```
See above.

alculquicondor commented on pkg/apis/batch/types.go:
```diff
@@ -31,7 +35,17 @@ import (
-const JobTrackingFinalizer = "batch.kubernetes.io/job-tracking"
+const JobTrackingFinalizer = LabelPrefix + "job-tracking"
+
+// LegacyJobName and LegacyControllerUid are legacy labels that were set using unprefixed labels.
+const LegacyJobName = "job-name"
+const LegacyControllerUid = "controller-uid"
```
If we don't want people to use the variables, we shouldn't have them exposed, even if they are named Legacy. You could keep them private, but that doesn't save us much.

kannon92 replied:
Not sure what your suggestion is here.  I expose them because I need to use them to generate the labels but I mostly wanted to mark these so when/if we remove these labels eventually it is easy to see which ones are the ones to remove.

alculquicondor commented on pkg/apis/batch/types.go:
```diff
@@ -22,6 +22,10 @@ import (
 
+// Unprefixed labels are reserved for end-users so we will add a batch.kubernetes.io to designate these labels as official Kubernetes labels.
+// See https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#label-selector-and-annotation-conventions
+const LabelPrefix = "batch.kubernetes.io/"
```
This doesn't need to be exported.
```suggestion
const labelPrefix = "batch.kubernetes.io/"
```

alculquicondor commented on pkg/registry/batch/job/strategy.go:
```diff
@@ -236,8 +243,12 @@ func generateSelector(obj *batch.Job) {
-		obj.Spec.Selector.MatchLabels["controller-uid"] = string(obj.ObjectMeta.UID)
+
+	// TODO: when job labels have both legacy labels and the official Kubernetes ones we should replace this selector to use batch.ControllerUid
+	for _, value := range []string{batch.LegacyControllerUid} {
```
We don't need the for.

But, as discussed in the other thread, we can probably just use the new labels already.

alculquicondor commented on staging/src/k8s.io/api/batch/v1/types.go:
```diff
@@ -23,8 +23,11 @@ import (
-	JobCompletionIndexAnnotation = "batch.kubernetes.io/job-completion-index"
+	// All Kubernetes labels need to be prefixed with Kubernetes to distinguish them from end-user labels
+	// More info: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#label-selector-and-annotation-conventions
+	LabelPrefix = "batch.kubernetes.io/"
```
same remarks here about exported variables.

kannon92 commented:
I addressed the comments that I was able too.  Please review when you get a chance.  

I discovered a bug in github where I can only request a review from one person..

alculquicondor commented on pkg/registry/batch/job/strategy.go:
```diff
@@ -236,8 +243,9 @@ func generateSelector(obj *batch.Job) {
-		obj.Spec.Selector.MatchLabels["controller-uid"] = string(obj.ObjectMeta.UID)
+
+	if _, found := obj.Spec.Selector.MatchLabels[batch.ControllerUid]; !found {
+		obj.Spec.Selector.MatchLabels[batch.ControllerUid] = string(obj.ObjectMeta.UID)
```
So the name was never actually used in the selector?

I wonder if the name label is relic from a time when we use to match on job name instead of UID.

kannon92 replied:
https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/batch/job/strategy.go#L211

Maybe.  This is the only comment I can find that explains the rational.

kannon92 commented on pkg/apis/batch/validation/validation.go:
```diff
@@ -99,12 +99,16 @@ func ValidateGeneratedSelector(obj *batch.Job) field.ErrorList {
 	// the user added coflicting labels.  Validate that the expected
 	// generated ones are there.
+	allErrs = append(allErrs, apivalidation.ValidateHasLabel(obj.Spec.Template.ObjectMeta, field.NewPath("spec").Child("template").Child("metadata"), batch.LegacyControllerUid, string(obj.UID))...)
+	allErrs = append(allErrs, apivalidation.ValidateHasLabel(obj.Spec.Template.ObjectMeta, field.NewPath("spec").Child("template").Child("metadata"), batch.LegacyJobName, string(obj.Name))...)
```
Ah. I misunderstood your ask.  I will add that as an optional validation.

kannon92 replied:
Alright I added an optional validation.

kannon92 commented on staging/src/k8s.io/api/batch/v1/types.go:
```diff
@@ -34,7 +37,12 @@ const (
+	// Historically the job controller uses unprefixed labels for job-name and controller-uid and
+	// Kubernetes continutes to recognize those unprefixed labels for consistency.
+	JobName       = labelPrefix + "job-name"
+	ControllerUid = labelPrefix + "controller-uid"
```
Done.

kannon92 commented:
/retest

kannon92 commented:
/retest

alculquicondor commented:
The release note says:

```
Fixed a bug where Kubernetes used unprefixed labels
```

This could be confusing to users. There was no bug, in the sense that everything was working. k8s originally didn't prefix many things. We have since realized that it's not a good idea, but that doesn't make it a bug :)

alculquicondor commented on pkg/apis/batch/types.go:
```diff
@@ -31,7 +35,17 @@ import (
+const LegacyControllerUid = "controller-uid"
+
+// JobName is a user friendly way to refer to jobs and is set in the labels for jobs.
+const JobName = labelPrefix + LegacyJobName
```
```suggestion
const JobNameLabel = labelPrefix + LegacyJobName
```

alculquicondor replied:
same for versioned package

kannon92 replied:
done.

alculquicondor commented on pkg/apis/batch/types.go:
```diff
@@ -31,7 +35,17 @@ import (
+const JobName = labelPrefix + LegacyJobName
+
+// Controller UID is used for selectors and labels for jobs
+const ControllerUid = labelPrefix + LegacyControllerUid
```
```suggestion
const ControllerUidLabel = labelPrefix + LegacyControllerUid
```

kannon92 replied:
done.

alculquicondor commented on pkg/apis/batch/validation/validation_test.go:
```diff
@@ -828,7 +854,7 @@ func TestValidateJob(t *testing.T) {
 	for k, v := range errorCases {
 		t.Run(k, func(t *testing.T) {
-			errs := ValidateJob(&v, JobValidationOptions{})
+			errs := ValidateJob(&v, JobValidationOptions{AllowBatchLabel: true})
```
the options should be part of the testcase

kannon92 replied:
I made this change but it was a pretty large one.  I think its good long term.

alculquicondor commented on pkg/registry/batch/job/strategy.go:
```diff
@@ -185,6 +186,11 @@ func validationOptionsForJob(newJob, oldJob *batch.Job) batchvalidation.JobValid
 
+		// Validation should not fail jobs if they don't have the new labels.
+		// This can be deprecated once we can guarantee both labels exist (1.29 at least)
+		_, validJobName := oldJob.Spec.Template.Labels[batch.JobName]
```
```suggestion
		_, hadJobName := oldJob.Spec.Template.Labels[batch.JobName]
```

alculquicondor commented:
```
The legacy labels `job-name` and `controller-uid` are still added for compatibility.
```

kannon92 commented:
/label api-review

alculquicondor commented on pkg/apis/batch/validation/validation_test.go:
```diff
@@ -245,590 +303,732 @@ func TestValidateJob(t *testing.T) {
+					Selector: &metav1.LabelSelector{
+						MatchLabels: map[string]string{"a": "b"},
+					},
+					Template: validPodTemplateSpecForGenerated,
```
also another case where the template is missing the uid label

kannon92 replied:
done.

alculquicondor commented:
/lgtm

alculquicondor commented:
Please don't forget to update the release note as suggested above https://github.com/kubernetes/kubernetes/pull/114930#issuecomment-1442029238

thockin commented on pkg/registry/batch/job/strategy.go:
```diff
@@ -237,8 +251,9 @@ func generateSelector(obj *batch.Job) {
-	if _, found := obj.Spec.Selector.MatchLabels["controller-uid"]; !found {
-		obj.Spec.Selector.MatchLabels["controller-uid"] = string(obj.ObjectMeta.UID)
+
+	if _, found := obj.Spec.Selector.MatchLabels[batch.ControllerUidLabel]; !found {
```
just to be sure I follow - existing Jobs, which have the "naked" label also have a selector which selects that label, so there's no risk of abandonment here, right?  New instances of Job will have the new label and the new selector, IIUC?

kannon92 replied:
From what @alculquicondor and I discussed I believe this is the case.  

We don't change selector or labels on updates.

alculquicondor replied:
That is correct, existing jobs are unaffected. New jobs will have the new selector.

thockin commented on pkg/apis/batch/validation/validation.go:
```diff
@@ -603,4 +611,6 @@ type JobValidationOptions struct {
 	// Allow elastic indexed jobs
 	AllowElasticIndexedJobs bool
+	// Allow Job to have the label on batch.kubernetes.io/job-name and batch.kubernetes.io/controller-uid
+	AllowBatchLabel bool
```
Here you call it "allow" implying it's optional, but `ValidateGeneratedSelector` is really _requiring_ values.

Can we rename this to `RequirePrefixedLabels` or something like that?

kannon92 replied:
done.

thockin reviewed (commented):
Problem 1:  `Validate` calls `generateSelector`  which adds labels.  This is wrong - `Validate` is not supposed to modify objects!  This should be done from `PrepareFor*` hooks instead.  A side-effect of this is that any update operation will set these labels, which seems like a benefit but is hidden in a place I would NEVER think to look.

I would retool all of this.  `PrepareForCreate` should call `generateSelectorIfNeeded`, which checks the `manual` flag and sets the labels.  It's not clear to me whether that should happen from `PrepareForUpdate` or not.  It's easy, but a bit unusual (to set unrelated fields) and low-value (I think).

Problem 2: `Validate` says `if job.Spec.ManualSelector == nil || *job.Spec.ManualSelector == false` - this is wrong.  That field should be recognized as `nil` in pkg/apis/batch/v1/defaults.go and assigned a value.  You should not be nil-checking fields all over the place and inferring their default values.  It seems innocuous on a bool, but imagine an int or string field - you'd have the default value sprinkled all over the codebase.

kannon92 commented:
> Problem 1: `Validate` calls `generateSelector` which adds labels. This is wrong - `Validate` is not supposed to modify objects! This should be done from `PrepareFor*` hooks instead. A side-effect of this is that any update operation will set these labels, which seems like a benefit but is hidden in a place I would NEVER think to look.
> 
> I would retool all of this. `PrepareForCreate` should call `generateSelectorIfNeeded`, which checks the `manual` flag and sets the labels. It's not clear to me whether that should happen from `PrepareForUpdate` or not. It's easy, but a bit unusual (to set unrelated fields) and low-value (I think).
> 
> Problem 2: `Validate` says `if job.Spec.ManualSelector == nil || *job.Spec.ManualSelector == false` - this is wrong. That field should be recognized as `nil` in pkg/apis/batch/v1/defaults.go and assigned a value. You should not be nil-checking fields all over the place and inferring their default values. It seems innocuous on a bool, but imagine an int or string field - you'd have the default value sprinkled all over the codebase.

@thockin These are all good points.  Could I take this comment and create an issue and solve that in a followup?

thockin commented:
> Could I take this comment and create an issue and solve that in a followup

Yes, of course :)

kannon92 commented on pkg/apis/batch/types.go:
```diff
@@ -22,16 +22,28 @@ import (
-// always track jobs using finalizers.
-const JobTrackingFinalizer = "batch.kubernetes.io/job-tracking"
+const (
+	// Unprefixed labels are reserved for end-users so we will add a batch.kubernetes.io to designate these labels as official Kubernetes labels.
```
done.

thockin commented:
gofmt complaining about trailing whitespace I think

kannon92 commented:
@alculquicondor and @thockin are there any outstanding items in this PR?  I believe I have addressed all outstanding issues but please let me know if I missed anything.  I had to resolve a merge conflict so I lost @alculquicondor lgtm.

thockin reviewed (commented):
https://github.com/kubernetes/kubernetes/pull/114930/files#r1119117149

is my only outstanding

kannon92 commented:
/retest

alculquicondor commented on pkg/apis/batch/validation/validation.go:
```diff
@@ -75,7 +75,7 @@ var (
 // TODO: generalize for other controller objects that will follow the same pattern, such as ReplicaSet and DaemonSet, and
 // move to new location.  Replace batch.Job with an interface.
-func ValidateGeneratedSelector(obj *batch.Job) field.ErrorList {
+func ValidateGeneratedSelector(obj *batch.Job, validateBatchLabels bool) field.ErrorList {
```
this probably doesn't need to be exported

kannon92 replied:
updated.

alculquicondor commented on pkg/registry/batch/job/strategy.go:
```diff
@@ -184,6 +185,12 @@ func validationOptionsForJob(newJob, oldJob *batch.Job) batchvalidation.JobValid
 		opts.AllowMutableSchedulingDirectives = suspended && notStarted
+
+		// Validation should not fail jobs if they don't have the new labels.
+		// This can be deprecated once we can guarantee both labels exist (1.29 at least)
```
```suggestion
		// This can be removed once we have high confidence that both labels exist (1.30 at least)
```

kannon92 replied:
updated.

kannon92 commented:
> https://github.com/kubernetes/kubernetes/pull/114930/files#r1119117149
> 
> is my only outstanding

Thanks for calling that out.  I sometimes miss items in gitub.  We addressed this in a separate PR so I think its worth a final look.

kannon92 commented:
@alculquicondor and @thockin can you look this over when you get a chance?

thockin commented:
Thanks!

/lgtm
/approve

thockin commented:
/retest
